In [51]:
def train_imat(method):
    method = str(method)
    !time java -jar ../../lib/RankLib-2.1-patched.jar -train ../../data/task3_train.txt -tts 0.2 -ranker {method} -metric2t NDCG@20 -sparse -silent

In [52]:
methods = [(0, "MART"), (1, "RankNet"), (2, "RankBoost"), (3, "AdaRank"), (4, "Coordinate Ascent"), (6, "LambdaMART"), (7, "ListNet"), (8, "Random Forests")]

In [53]:
for method, name in methods:
    print("Method: " + name)
    train_imat(method)

Method: MART

[+] General Parameters:
Training data:	../../data/task3_train.txt
Train-Test split: 0.2
Feature vector representation: Sparse.
Ranking method:	MART
Feature description file:	Unspecified. All features will be used.
Train metric:	NDCG@20
Test metric:	NDCG@20
Feature normalization: No

[+] MART's Parameters:

Reading feature file [../../data/task3_train.txt]... [Done.]            
(7033 ranked lists, 75057 entries read)
NDCG@20 on test data: 0.7415
351.80user 0.96system 4:34.93elapsed 128%CPU (0avgtext+0avgdata 631952maxresident)k
0inputs+576outputs (0major+169759minor)pagefaults 0swaps
Method: RankNet

[+] General Parameters:
Training data:	../../data/task3_train.txt
Train-Test split: 0.2
Feature vector representation: Sparse.
Ranking method:	RankNet
Feature description file:	Unspecified. All features will be used.
Train metric:	NDCG@20
Test metric:	NDCG@20
Feature normalization: No

[+] RankNet's Parameters:

Reading feature file [../../data/task3_train.txt]... [Done.]    

b[137]    | 0.7575    | 
b[138]    | 0.7483    | 
b[139]    | 0.726     | 
b[140]    | 0.7414    | 
b[141]    | 0.726     | 
b[142]    | 0.7416    | 
b[143]    | 0.745     | 
b[144]    | 0.74      | 
b[145]    | 0.7258    | 
b[146]    | 0.7376    | 
b[147]    | 0.7479    | 
b[148]    | 0.7459    | 
b[149]    | 0.7345    | 
b[150]    | 0.747     | 
b[151]    | 0.7474    | 
b[152]    | 0.7413    | 
b[153]    | 0.7519    | 
b[154]    | 0.7384    | 
b[155]    | 0.7396    | 
b[156]    | 0.7515    | 
b[157]    | 0.7527    | 
b[158]    | 0.7399    | 
b[159]    | 0.7432    | 
b[160]    | 0.7327    | 
b[161]    | 0.7395    | 
b[162]    | 0.7336    | 
b[163]    | 0.7353    | 
b[164]    | 0.7198    | 
b[165]    | 0.7408    | 
b[166]    | 0.745     | 
b[167]    | 0.7291    | 
b[168]    | 0.743     | 
b[169]    | 0.7209    | 
b[170]    | 0.7412    | 
b[171]    | 0.7303    | 
b[172]    | 0.7269    | 
b[173]    | 0.7418    | 
b[174]    | 0.7462    | 
b[175]    | 0.7401    | 
b[176]    | 0.7435    | 


Method: MART  
NDCG@20 on test data: 0.7415  
351.80  

Method: RankNet  
NDCG@20 on test data: 0.6785  
189.3  

Method: RankBoost  
NDCG@20 on test data: 0.7184  
359.46  

Method: AdaRank  
NDCG@20 on test data: 0.7177  
84.87  

Method: Coordinate Ascent  
NDCG@20 on test data: 0.73  
3704.80  

Method: LambdaMART  
NDCG@20 on test data: 0.7418  
399.58  

Method: ListNet  
NDCG@20 on test data: 0.7247  
665.07  

Method: Random Forests  
NDCG@20 on test data: 0.7447  
917.35

Интересны MART и Random Forests, если пытаться получить лучший результат

# MART

In [61]:
!time java -jar ../../lib/RankLib-2.1-patched.jar -train ../../data/task3_train.txt -tts 0.2 -ranker 0 -tree 3000 -leaf 100 -metric2t NDCG@20 -sparse -silent


[+] General Parameters:
Training data:	../../data/task3_train.txt
Train-Test split: 0.2
Feature vector representation: Sparse.
Ranking method:	MART
Feature description file:	Unspecified. All features will be used.
Train metric:	NDCG@20
Test metric:	NDCG@20
Feature normalization: No

[+] MART's Parameters:

Reading feature file [../../data/task3_train.txt]... [Done.]            
(7033 ranked lists, 75057 entries read)
NDCG@20 on test data: 0.7389
4451.95user 26.62system 1:04:07elapsed 116%CPU (0avgtext+0avgdata 932280maxresident)k
228696inputs+7496outputs (2major+11271291minor)pagefaults 0swaps


In [62]:
!time java -jar ../../lib/RankLib-2.1-patched.jar -train ../../data/task3_train.txt -tts 0.2 -ranker 0 -tree 3000 -leaf 10 -metric2t NDCG@20 -sparse -silent


[+] General Parameters:
Training data:	../../data/task3_train.txt
Train-Test split: 0.2
Feature vector representation: Sparse.
Ranking method:	MART
Feature description file:	Unspecified. All features will be used.
Train metric:	NDCG@20
Test metric:	NDCG@20
Feature normalization: No

[+] MART's Parameters:

Reading feature file [../../data/task3_train.txt]... [Done.]            
(7033 ranked lists, 75057 entries read)
NDCG@20 on test data: 0.7355
1004.63user 1.96system 12:55.99elapsed 129%CPU (0avgtext+0avgdata 594688maxresident)k
0inputs+1576outputs (0major+159373minor)pagefaults 0swaps


In [65]:
!time java -jar ../../lib/RankLib-2.1-patched.jar -train ../../data/task3_train.txt -tts 0.2 -ranker 0 -tree 500 -leaf 10 -metric2t NDCG@20 -sparse -silent


[+] General Parameters:
Training data:	../../data/task3_train.txt
Train-Test split: 0.2
Feature vector representation: Sparse.
Ranking method:	MART
Feature description file:	Unspecified. All features will be used.
Train metric:	NDCG@20
Test metric:	NDCG@20
Feature normalization: No

[+] MART's Parameters:

Reading feature file [../../data/task3_train.txt]... [Done.]            
(7033 ranked lists, 75057 entries read)
NDCG@20 on test data: 0.7428
177.46user 0.64system 2:16.16elapsed 130%CPU (0avgtext+0avgdata 634316maxresident)k
0inputs+288outputs (0major+170177minor)pagefaults 0swaps


In [67]:
!time java -jar ../../lib/RankLib-2.1-patched.jar -train ../../data/task3_train.txt -tts 0.2 -ranker 0 -tree 500 -leaf 20 -metric2t NDCG@20 -sparse -silent


[+] General Parameters:
Training data:	../../data/task3_train.txt
Train-Test split: 0.2
Feature vector representation: Sparse.
Ranking method:	MART
Feature description file:	Unspecified. All features will be used.
Train metric:	NDCG@20
Test metric:	NDCG@20
Feature normalization: No

[+] MART's Parameters:

Reading feature file [../../data/task3_train.txt]... [Done.]            
(7033 ranked lists, 75057 entries read)
NDCG@20 on test data: 0.742
276.14user 0.76system 3:52.51elapsed 119%CPU (0avgtext+0avgdata 702884maxresident)k
0inputs+320outputs (0major+190174minor)pagefaults 0swaps


# Random Forests

In [63]:
!time java -jar ../../lib/RankLib-2.1-patched.jar -train ../../data/task3_train.txt -tts 0.2 -ranker 8 -bag 300 -leaf 100 -metric2t NDCG@20 -sparse -silent


[+] General Parameters:
Training data:	../../data/task3_train.txt
Train-Test split: 0.2
Feature vector representation: Sparse.
Ranking method:	Random Forests
Feature description file:	Unspecified. All features will be used.
Train metric:	NDCG@20
Test metric:	NDCG@20
Feature normalization: No

[+] Random Forests's Parameters:

Reading feature file [../../data/task3_train.txt]... [Done.]            
(7033 ranked lists, 75057 entries read)
b[1]      | 0.7346    | 
b[2]      | 0.7221    | 
b[3]      | 0.7276    | 
b[4]      | 0.729     | 
b[5]      | 0.7353    | 
b[6]      | 0.7276    | 
b[7]      | 0.7396    | 
b[8]      | 0.733     | 
b[9]      | 0.7113    | 
b[10]     | 0.7182    | 
b[11]     | 0.7348    | 
b[12]     | 0.736     | 
b[13]     | 0.736     | 
b[14]     | 0.7321    | 
b[15]     | 0.7505    | 
b[16]     | 0.7574    | 
b[17]     | 0.7338    | 
b[18]     | 0.7314    | 
b[19]     | 0.7416    | 
b[20]     | 0.7375    | 
b[21]     | 0.7446    | 
b[22]     | 0.7328    | 
b[23]   

In [64]:
!time java -jar ../../lib/RankLib-2.1-patched.jar -train ../../data/task3_train.txt -tts 0.2 -ranker 8 -bag 300 -leaf 10 -metric2t NDCG@20 -sparse -silent


[+] General Parameters:
Training data:	../../data/task3_train.txt
Train-Test split: 0.2
Feature vector representation: Sparse.
Ranking method:	Random Forests
Feature description file:	Unspecified. All features will be used.
Train metric:	NDCG@20
Test metric:	NDCG@20
Feature normalization: No

[+] Random Forests's Parameters:

Reading feature file [../../data/task3_train.txt]... [Done.]            
(7033 ranked lists, 75057 entries read)
b[1]      | 0.6885    | 
b[2]      | 0.7005    | 
b[3]      | 0.7048    | 
b[4]      | 0.6928    | 
b[5]      | 0.6673    | 
b[6]      | 0.6898    | 
b[7]      | 0.6897    | 
b[8]      | 0.693     | 
b[9]      | 0.6906    | 
b[10]     | 0.685     | 
b[11]     | 0.6892    | 
b[12]     | 0.6813    | 
b[13]     | 0.7011    | 
b[14]     | 0.6845    | 
b[15]     | 0.6905    | 
b[16]     | 0.6883    | 
b[17]     | 0.689     | 
b[18]     | 0.6821    | 
b[19]     | 0.6966    | 
b[20]     | 0.703     | 
b[21]     | 0.6856    | 
b[22]     | 0.6831    | 
b[23]   

In [66]:
!time java -jar ../../lib/RankLib-2.1-patched.jar -train ../../data/task3_train.txt -tts 0.2 -ranker 8 -bag 1000 -leaf 20 -metric2t NDCG@20 -sparse -silent


[+] General Parameters:
Training data:	../../data/task3_train.txt
Train-Test split: 0.2
Feature vector representation: Sparse.
Ranking method:	Random Forests
Feature description file:	Unspecified. All features will be used.
Train metric:	NDCG@20
Test metric:	NDCG@20
Feature normalization: No

[+] Random Forests's Parameters:

Reading feature file [../../data/task3_train.txt]... [Done.]            
(7033 ranked lists, 75057 entries read)
b[1]      | 0.7055    | 
b[2]      | 0.7087    | 
b[3]      | 0.6953    | 
b[4]      | 0.6842    | 
b[5]      | 0.7084    | 
b[6]      | 0.7057    | 
b[7]      | 0.672     | 
b[8]      | 0.7054    | 
b[9]      | 0.7035    | 
b[10]     | 0.6827    | 
b[11]     | 0.6982    | 
b[12]     | 0.7006    | 
b[13]     | 0.7053    | 
b[14]     | 0.6921    | 
b[15]     | 0.6988    | 
b[16]     | 0.7072    | 
b[17]     | 0.7107    | 
b[18]     | 0.6896    | 
b[19]     | 0.7006    | 
b[20]     | 0.6944    | 
b[21]     | 0.6894    | 
b[22]     | 0.7104    | 
b[23]   

b[312]    | 0.7102    | 
b[313]    | 0.7024    | 
b[314]    | 0.6975    | 
b[315]    | 0.7071    | 
b[316]    | 0.7019    | 
b[317]    | 0.6889    | 
b[318]    | 0.7014    | 
b[319]    | 0.6989    | 
b[320]    | 0.6925    | 
b[321]    | 0.6912    | 
b[322]    | 0.6944    | 
b[323]    | 0.6929    | 
b[324]    | 0.7028    | 
b[325]    | 0.6863    | 
b[326]    | 0.7107    | 
b[327]    | 0.7129    | 
b[328]    | 0.696     | 
b[329]    | 0.7006    | 
b[330]    | 0.6984    | 
b[331]    | 0.6952    | 
b[332]    | 0.6945    | 
b[333]    | 0.6927    | 
b[334]    | 0.724     | 
b[335]    | 0.7133    | 
b[336]    | 0.6905    | 
b[337]    | 0.6899    | 
b[338]    | 0.6888    | 
b[339]    | 0.714     | 
b[340]    | 0.7106    | 
b[341]    | 0.7037    | 
b[342]    | 0.7067    | 
b[343]    | 0.7029    | 
b[344]    | 0.7076    | 
b[345]    | 0.7088    | 
b[346]    | 0.6939    | 
b[347]    | 0.6981    | 
b[348]    | 0.6961    | 
b[349]    | 0.7025    | 
b[350]    | 0.7012    | 
b[351]    | 0.7153    | 


b[640]    | 0.7085    | 
b[641]    | 0.6849    | 
b[642]    | 0.7051    | 
b[643]    | 0.7113    | 
b[644]    | 0.7012    | 
b[645]    | 0.7085    | 
b[646]    | 0.6897    | 
b[647]    | 0.7076    | 
b[648]    | 0.7052    | 
b[649]    | 0.7003    | 
b[650]    | 0.7042    | 
b[651]    | 0.6968    | 
b[652]    | 0.7037    | 
b[653]    | 0.6993    | 
b[654]    | 0.6953    | 
b[655]    | 0.6861    | 
b[656]    | 0.697     | 
b[657]    | 0.7035    | 
b[658]    | 0.7218    | 
b[659]    | 0.7068    | 
b[660]    | 0.7165    | 
b[661]    | 0.6976    | 
b[662]    | 0.7031    | 
b[663]    | 0.7069    | 
b[664]    | 0.6811    | 
b[665]    | 0.7024    | 
b[666]    | 0.6989    | 
b[667]    | 0.6884    | 
b[668]    | 0.7039    | 
b[669]    | 0.6918    | 
b[670]    | 0.7191    | 
b[671]    | 0.6863    | 
b[672]    | 0.7005    | 
b[673]    | 0.7076    | 
b[674]    | 0.7035    | 
b[675]    | 0.7132    | 
b[676]    | 0.7063    | 
b[677]    | 0.7023    | 
b[678]    | 0.702     | 
b[679]    | 0.683     | 


b[968]    | 0.6951    | 
b[969]    | 0.6969    | 
b[970]    | 0.7041    | 
b[971]    | 0.7141    | 
b[972]    | 0.6967    | 
b[973]    | 0.7127    | 
b[974]    | 0.6987    | 
b[975]    | 0.7032    | 
b[976]    | 0.697     | 
b[977]    | 0.703     | 
b[978]    | 0.701     | 
b[979]    | 0.7031    | 
b[980]    | 0.7064    | 
b[981]    | 0.6885    | 
b[982]    | 0.7015    | 
b[983]    | 0.71      | 
b[984]    | 0.7199    | 
b[985]    | 0.7084    | 
b[986]    | 0.7053    | 
b[987]    | 0.6924    | 
b[988]    | 0.6962    | 
b[989]    | 0.6951    | 
b[990]    | 0.7097    | 
b[991]    | 0.6885    | 
b[992]    | 0.7007    | 
b[993]    | 0.7122    | 
b[994]    | 0.6952    | 
b[995]    | 0.7043    | 
b[996]    | 0.7023    | 
b[997]    | 0.7112    | 
b[998]    | 0.7021    | 
b[999]    | 0.6997    | 
b[1000]   | 0.7065    | 
------------------------------------
Finished sucessfully.
NDCG@20 on training data: 0.7352
------------------------------------
NDCG@20 on test data: 0.7376
2699.91user 36.40